pkg/controllers/provisioner/v1alpha1/allocation/greedyallocator.go

```diff
@@ -40,143 +39,82 @@ type GreedyAllocator struct {
 func (a *GreedyAllocator) Allocate(provisioner *v1alpha1.Provisioner, pods []*v1.Pod) error {
 	// 1. Separate pods into scheduling groups
 	groups := a.getSchedulingGroups(pods)
+	ctx := context.TODO() // TODO wire this in from reconcile loop
 
 	zap.S().Infof("Allocating %d pending pods from %d constraint groups", len(pods), len(groups))
 	// 2. Group pods into equally schedulable constraint group
-	for _, group := range groups {
-		nodes, err := a.CloudProvider.CapacityFor(&provisioner.Spec).Create(context.TODO(), group.Constraints)
+	for _, constraints := range groups {
+		packing, err := a.CloudProvider.CapacityFor(&provisioner.Spec).Create(ctx, constraints)
 		// TODO accumulate errors if one request fails.
 		if err != nil {
 			return fmt.Errorf("while creating capacity, %w", err)
 		}
-		if err := a.createNodesAndAssignPods(nodes, group.Pods); err != nil {
-			return fmt.Errorf("assigning pods to nodes err: %w", err)
+		for node, pods := range packing {
+			if err := a.bind(ctx, node, pods); err != nil {
+				// TODO accumulate errors if one request fails.
+				return fmt.Errorf("binding pods to node, %w", err)
+			}
 		}
 	}
 	return nil
 }
 
-func (a *GreedyAllocator) createNodesAndAssignPods(nodes []*v1.Node, pods []*v1.Pod) error {
-
-	// Currently we are assigning each pod per node.
-	// Create node object in the cluster
-	// score nodes and pods from bigger to lower in the list
-	remainingPods := make([]*v1.Pod, len(pods))
-	copy(remainingPods, pods)
-	for _, node := range nodes {
-		err := a.createNodeObject(node)
-		if err != nil {
-			return fmt.Errorf("creating node object %w", err)
-		}
-		remainingPods, err = a.assignPodsToNodes(node, remainingPods)
-		if err != nil {
-			return fmt.Errorf("update pod spec err: %w", err)
-		}
-	}
-	if len(remainingPods) > 0 {
-		// this should not happen
-		return fmt.Errorf("unable to assign %d pods to %d nodes", len(remainingPods), len(nodes))
+func (a *GreedyAllocator) bind(ctx context.Context, node *v1.Node, pods []*v1.Pod) error {
+	// 1. Create node object
+	if _, err := a.CoreV1Client.Nodes().Create(ctx, node, metav1.CreateOptions{}); err != nil {
+		return fmt.Errorf("creating node %s, %w", node.Name, err)
 	}
-	zap.S().Infof("Successfully assigned %d pods to %d node ", len(pods), len(nodes))
-	return nil
-}
-
-func (a *GreedyAllocator) assignPodsToNodes(node *v1.Node, pods []*v1.Pod) ([]*v1.Pod, error) {
-
-	remainingPods := make([]*v1.Pod, 0)
+	// 2. Bind all pods to node
 	for _, pod := range pods {
-		if !canFitPodOnNode(node, pod) {
-			remainingPods = append(remainingPods, pod)
-			continue
-		}
-		if err := a.bindPodToNode(node, pod); err != nil {
-			return nil, fmt.Errorf("binding pod to node failed err %w", err)
+		if err := a.CoreV1Client.Pods(pod.Namespace).Bind(ctx, &v1.Binding{
+			TypeMeta: pod.TypeMeta, ObjectMeta: pod.ObjectMeta,
```

Comment:
Shall we separate the struct fields in separate lines?

Reply:
Yeah, good call.